### DISCUSSION TITLE
How to implement vue-logger-plugin properly?

### DISCUSSION THREAD

**Discussion by rowild** [quasarframework/quasar#13667] [General - Components / Directives / etc] [answered]
The https://github.com/dev-tavern/vue-logger-plugin recommends to create a new file and import that one, if there is need for custom before and/or after hooks. This is the example from the docu:

```js
import { createLogger } from 'vue-logger-plugin'

// create logger with options
const logger = createLogger({
  enabled: true,
  level: 'debug',
  beforeHooks: [ ... ],
  afterHooks: [ ... ]
})

export default logger
```

Reading about boot files and the cases when they are supposed to be used or not, I decided to try to go with a boot file that looks like this (based on the `axios` example in the quasar docu):

```js
import { boot } from 'quasar/wrappers'
import { createLogger } from 'vue-logger-plugin'

export default boot(({ app }) => {
  console.log('invoked boot function with logger');

  // create logger with options
  const logger = createLogger({
    enabled: true,
    level: 'debug',
    beforeHooks: [],
    afterHooks: [],
  })

  // app.config.globalProperties.$log = logger
  // ^ ^ ^ this will allow you to use this.$logger (for Vue Options API form)
  //       so you won't necessarily have to import logger in each vue file

  app.use(logger)
})
```

The I add 'logger' to the `quasar.config.js`'s boot array:

```js
module.exports = configure(function (ctx) {
  boot: [ 'logger', 'axios']
}
```

However, I admit to have to problems now:

1. I still do not know how to use `logger` now. Similar to the `axios` example I would expect that a simple `logger("blasbla")` would suffice, but instead a `error  'logger' is not defined ` is thrown.

2. Is implementing a logger via quasar's boot option really the correct/recommended way? The docu says sth [freely cited] "use it only if a script has to run before Vue".

**Comment by yusufkandemir** [comment, marked as answer]
1. vue-logger-plugin and Axios are different libraries, so they are not initialized or used the same way. See the [usage section of vue-logger-plugin](https://github.com/dev-tavern/vue-logger-plugin#usage) for how to use it.

Composition API:
```js
import { useLogger } from 'vue-logger-plugin'

// Inside <script setup> or `setup()`
const log = useLogger()
log.info('Setting up MyComponent...')
```
Options API:
```js
// ...
export default defineComponent({
  // ...
  created() {
    this.$log.info('Setting up MyComponent...')
  }
}
```

2. Yes, it's the recommended way, and in this instance, it's also the only way to do it. Because you need to access the `app`.
https://quasar.dev/quasar-cli-vite/boot-files#when-to-use-boot-files
> Boot files fulfill one special purpose: they run code before the App’s Vue root component is instantiated while giving you access to certain variables, which is required **if you need to initialize a library**, interfere with Vue Router, inject Vue prototype or inject the root instance of the Vue app.

**Reply by rowild** [reply]
Thank you for your feedback, @yusufkandemir ! Especially for confirming my 2nd question! :-)

But the regular use of the plugin - via `useLogger` - does not respect the "Advanced Usage" (See docu => Vue Plugin Install), for which I am seeking a solution. Or, in case I misunderstand, where would I setup custom hooks?

**Reply by yusufkandemir** [reply]
vue-logger-plugin suggests a separate file to keep main.js simpler, but since boot files already allow you to separate each library or concern into multiple files, technically you already done that "separate file" step. So, you don't need to do anything special, export or import anything. Just set `beforeHooks`, `afterHooks`, etc. in the boot file you've sent, inside `createLogger` call.

```js
import { boot } from 'quasar/wrappers'
import { createLogger, StringifyAndParseObjectsHook } from 'vue-logger-plugin'

export default boot(({ app }) => {
  const logger = createLogger({
    enabled: true,
    level: 'debug',
    beforeHooks: [StringifyAndParseObjectsHook],
    afterHooks: [],
  })

  app.use(logger)
})
```

**Reply by rowild** [reply]
So that means that, even though I create a boot file for `vue-logger-plugin`, I still have to import and instantiate it into every file, where I want to use it via 

```js
import { useLogger } from 'vue-logger-plugin'
const log = useLogger()
```

That seems like "double the trouble" to me... My thinking is is that as soon as there is a `app.use(logger)`, `logger` is available everywhere. Isn't that one of the points of boot files? To make things available everywhere? Like the axios example? I am confused... why does axios work like that but logger not?

Thank you again! :-)

**Reply by yusufkandemir** [reply]
It's not "double the trouble", it's how JS and Vue work. Same with `useQuasar`, `useRouter`, `useStore`, etc. Axios is not a Vue library, it doesn't use provide/inject, and the initialization is not mandatory either. It just works differently. To explain why they are fundamentally different, I would have to explain how JS, global state, provide/inject of Vue, etc. works as well.

Looking at the [vue-logger-plugin's implementation](https://github.com/dev-tavern/vue-logger-plugin/blob/develop/src/logger.ts), the initialization step is to just enable provide/inject, and expose `$log` and `$logger`. So, aside from possible SSR concerns, you can directly expose the instance and use it in a similar way as you do with Axios.

```js
import { boot } from 'quasar/wrappers'
import { createLogger, StringifyAndParseObjectsHook } from 'vue-logger-plugin'

export const logger = createLogger({
  enabled: true,
  level: 'debug',
  beforeHooks: [StringifyAndParseObjectsHook],
  afterHooks: [],
})

// You can avoid this step if you will not use `useLogger`, `this.$log`, and `this.$logger`. You would also need to make sure `vue-logger-plugin` will not use a different implementation which requires this, in the future.
export default boot(({ app }) => {
  app.use(logger)
})
```

This way you can simplify this:
```js
import { useLogger } from 'vue-logger-plugin'

const log = useLogger()
logger.info('')
```
to this:
```js
import { logger } from 'boot/logger' // or whatever filename you are using

logger.info('')
```
I don't think it simplifies a lot of stuff, but sharing it to hopefully clear your confusion. This is not related to Quasar btw, it's just how some libraries work.

**Reply by rowild** [reply]
Wow! This is a fantastic explanation and showcase! Thank you very, very much, @yusufkandemir !
